### PR TITLE
Update/kafka implementations [WIP]

### PR DESCRIPTION
**Pull request by NicholasTurner23**
## Description

This PR add dynamic kafka producer and consumer to be used in the kafka integration.

## Related Issues
- [ ] JIRA cards:
  - [ ] OPS-291

## Summary by CodeRabbit

- **New Features**
	- Enhanced message delivery and error handling for Kafka interactions.
	- Introduced a method for selecting the least loaded partition for message sending.
	- Added functionality to update hourly data topics with device metadata.
	- Improved message publishing with optional keys and automatic partition selection.

- **Bug Fixes**
	- Improved logging for message delivery success and error handling.

- **Chores**
	- Updated package dependencies by removing `kafka-python` and adding `confluent-kafka` and `lz4`.

### CODE REVIEW

File: src/workflows/airqo_etl_utils/message_broker_utils.py [hunk 1]
```diff
@@ -132,3 +164,295 @@
...
+                selected_partition = (
+                    None if auto_partition else self.__get_least_loaded_partition()
+                )
+                self._send_message(producer, topic, None, message, selected_partition)
+
+                if not auto_partition:
+                    self.partition_loads[selected_partition] += 1
+
+        producer.close()
+
```
[the remark this answers is not in the document]
Comment (NicholasTurner23) [reply src/workflows/airqo_etl_utils/message_broker_utils.py:362]: The `close()` method also performs a flush if I am not mistaken.

File: src/workflows/airqo_etl_utils/message_broker_utils.py [hunk 2]
```diff
@@ -132,3 +164,295 @@
...
+                message_count += 1
+
+        except Exception as e:
+            logger.exception(f"Error while consuming messages from topic {topic}: {e}")
+
+        finally:
+            consumer.close()
+            logger.info(
+                f"Closed consumer. No more messages to consume from topic: {topic}"
+            )
```
[the remark this answers is not in the document]
Comment (NicholasTurner23) [reply src/workflows/airqo_etl_utils/message_broker_utils.py:458]: The consume method can be used for streaming where it runs indefinitely as well as consuming data once when triggered. At least that is the intention.

File: src/workflows/airqo_etl_utils/message_broker_utils.py [hunk 3]
```diff
@@ -132,3 +164,295 @@
...
+                    partition=partition,
+                    on_delivery=self.__on_delivery,
+                )
+            else:
+                producer.produce(
+                    topic=topic, key=key, value=message, on_delivery=self.__on_delivery
+                )
+        except Exception as e:
+            logger.exception(f"Error while sending message to topic {topic}: {e}")
+
```
[the remark this answers is not in the document]
Comment (NicholasTurner23) [reply src/workflows/airqo_etl_utils/message_broker_utils.py:302]: This is being done in the caller method.

**Review by Baalmart — COMMENTED** [review]
Thank you, @NicholasTurner23 , for this amazing contribution. I sincerely appreciate your efforts.

Regarding these proposed changes:

1. **Backward Compatibility:** Could you please provide insights on the potential impact of these modifications on existing functionality? Specifically:

- Are there any known issues or conflicts with currently deployed systems?
- Have compatibility tests been conducted with our production environment?

2. **Critical Security Systems:** Our organization has several critical security systems heavily reliant on Kafka. These systems need to remain operational throughout the ongoing work. Therefore, I would like to request:

- Confirmation that the proposed changes will not disrupt the current Kafka-based systems.
- Any mitigation strategies or rollback plans in place to ensure uninterrupted service during the transition period.

**Comment by NicholasTurner23** [comment]
> Thank you, @NicholasTurner23 , for this amazing contribution. I sincerely appreciate your efforts.
> 
> Regarding these proposed changes:
> 
>     1. **Backward Compatibility:** Could you please provide insights on the potential impact of these modifications on existing functionality? Specifically:
>  _At the moment, no impact on existing functionality. These changes are for the workflows application(s)._
>     2. **Critical Security Systems:** Our organization has several critical security systems heavily reliant on Kafka. These systems need to remain operational throughout the ongoing work. Therefore, I would like to request:
> _There is no interaction with other kafka systems. These changes only affect the workflows application(s)._

**Pull request merged**